<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>ANALYSIS OF SINGLE CELL RNA-SEQ DATA</title>
  <meta name="description" content="This is a minimal example of using the bookdown package to write a book. The output format for this example is bookdown::gitbook." />
  <meta name="generator" content="bookdown 0.16 and GitBook 2.6.7" />

  <meta property="og:title" content="ANALYSIS OF SINGLE CELL RNA-SEQ DATA" />
  <meta property="og:type" content="book" />
  <meta property="og:url" content="https://broadinstitute.github.io/2020_scWorkshop/" />
  <meta property="og:image" content="https://broadinstitute.github.io/2020_scWorkshop/images/cover.png" />
  <meta property="og:description" content="This is a minimal example of using the bookdown package to write a book. The output format for this example is bookdown::gitbook." />
  <meta name="github-repo" content="broadinstitute/2019_scWorkship" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="ANALYSIS OF SINGLE CELL RNA-SEQ DATA" />
  
  <meta name="twitter:description" content="This is a minimal example of using the bookdown package to write a book. The output format for this example is bookdown::gitbook." />
  <meta name="twitter:image" content="https://broadinstitute.github.io/2020_scWorkshop/images/cover.png" />

<meta name="author" content="Orr Ashenberg" />
<meta name="author" content="Dana Silverbush" />
<meta name="author" content="Kirk Gosik" />



  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  

<link rel="next" href="lab-template.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />











<style type="text/css">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">ANALYSIS OF SINGLE CELL RNA-SEQ DATA</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#course-overview"><i class="fa fa-check"></i><b>1.1</b> COURSE OVERVIEW</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#targeted-audience-assumed-background"><i class="fa fa-check"></i><b>1.2</b> TARGETED AUDIENCE &amp; ASSUMED BACKGROUND</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#course-format"><i class="fa fa-check"></i><b>1.3</b> COURSE FORMAT</a></li>
<li class="chapter" data-level="1.4" data-path="index.html"><a href="index.html#getting-started"><i class="fa fa-check"></i><b>1.4</b> Getting Started</a></li>
<li class="chapter" data-level="1.5" data-path="index.html"><a href="index.html#session-content"><i class="fa fa-check"></i><b>1.5</b> SESSION CONTENT</a><ul>
<li class="chapter" data-level="1.5.1" data-path="index.html"><a href="index.html#monday"><i class="fa fa-check"></i><b>1.5.1</b> Monday</a></li>
<li class="chapter" data-level="1.5.2" data-path="index.html"><a href="index.html#tuesday-classes-from-0800-to-1600"><i class="fa fa-check"></i><b>1.5.2</b> Tuesday – Classes from 08:00 to 16:00</a></li>
<li class="chapter" data-level="1.5.3" data-path="index.html"><a href="index.html#wednesday-classes-from-0800-to-1600"><i class="fa fa-check"></i><b>1.5.3</b> Wednesday – Classes from 08:00 to 16:00</a></li>
<li class="chapter" data-level="1.5.4" data-path="index.html"><a href="index.html#thursday-classes-from-0800-to-1600"><i class="fa fa-check"></i><b>1.5.4</b> Thursday – Classes from 08:00 to 16:00</a></li>
<li class="chapter" data-level="1.5.5" data-path="index.html"><a href="index.html#friday-classes-from-0800-to-1600"><i class="fa fa-check"></i><b>1.5.5</b> Friday – Classes from 08:00 to 16:00</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="lab-template.html"><a href="lab-template.html"><i class="fa fa-check"></i><b>2</b> Lab Template</a><ul>
<li class="chapter" data-level="2.1" data-path="lab-template.html"><a href="lab-template.html#template-example"><i class="fa fa-check"></i><b>2.1</b> Template Example</a></li>
<li class="chapter" data-level="2.2" data-path="lab-template.html"><a href="lab-template.html#make-object"><i class="fa fa-check"></i><b>2.2</b> Make Object</a></li>
<li class="chapter" data-level="2.3" data-path="lab-template.html"><a href="lab-template.html#process-object"><i class="fa fa-check"></i><b>2.3</b> Process Object</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="lecture-template.html"><a href="lecture-template.html"><i class="fa fa-check"></i><b>3</b> Lecture Template</a></li>
<li class="chapter" data-level="4" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html"><i class="fa fa-check"></i><b>4</b> Understanding Sequencing Raw Data</a><ul>
<li class="chapter" data-level="4.1" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#class-environment"><i class="fa fa-check"></i><b>4.1</b> Class Environment</a><ul>
<li class="chapter" data-level="4.1.1" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#getting-into-aws-instance"><i class="fa fa-check"></i><b>4.1.1</b> Getting into AWS Instance</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#shell-and-unix-commands"><i class="fa fa-check"></i><b>4.2</b> Shell and Unix commands</a><ul>
<li class="chapter" data-level="4.2.1" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#common-linux-commands"><i class="fa fa-check"></i><b>4.2.1</b> Common Linux Commands</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#file-formats"><i class="fa fa-check"></i><b>4.3</b> File formats</a><ul>
<li class="chapter" data-level="4.3.1" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#view-fastq-files"><i class="fa fa-check"></i><b>4.3.1</b> View FASTQ Files</a></li>
<li class="chapter" data-level="4.3.2" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#view-bam-files"><i class="fa fa-check"></i><b>4.3.2</b> View BAM Files</a></li>
</ul></li>
<li class="chapter" data-level="4.4" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#public-data-repositories"><i class="fa fa-check"></i><b>4.4</b> Public data repositories</a><ul>
<li class="chapter" data-level="4.4.1" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#cellranger10x"><i class="fa fa-check"></i><b>4.4.1</b> Cellranger/10x</a></li>
<li class="chapter" data-level="4.4.2" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#geo"><i class="fa fa-check"></i><b>4.4.2</b> GEO</a></li>
<li class="chapter" data-level="4.4.3" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#single-cell-portal"><i class="fa fa-check"></i><b>4.4.3</b> Single Cell Portal</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html"><i class="fa fa-check"></i><b>5</b> Introduction R/Bioconductor</a><ul>
<li class="chapter" data-level="5.1" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#start-environment"><i class="fa fa-check"></i><b>5.1</b> Start Environment</a></li>
<li class="chapter" data-level="5.2" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#installing-packages"><i class="fa fa-check"></i><b>5.2</b> Installing packages</a><ul>
<li class="chapter" data-level="5.2.1" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#cran"><i class="fa fa-check"></i><b>5.2.1</b> CRAN</a></li>
<li class="chapter" data-level="5.2.2" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#github"><i class="fa fa-check"></i><b>5.2.2</b> Github</a></li>
<li class="chapter" data-level="5.2.3" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#bioconductor"><i class="fa fa-check"></i><b>5.2.3</b> Bioconductor</a></li>
<li class="chapter" data-level="5.2.4" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#source"><i class="fa fa-check"></i><b>5.2.4</b> Source</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#installation-instructions"><i class="fa fa-check"></i><b>5.3</b> Installation instructions:</a><ul>
<li class="chapter" data-level="5.3.1" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#classestypes"><i class="fa fa-check"></i><b>5.3.1</b> Classes/Types</a></li>
<li class="chapter" data-level="5.3.2" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#data-structures"><i class="fa fa-check"></i><b>5.3.2</b> Data structures</a></li>
<li class="chapter" data-level="5.3.3" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#detour-to-s3s4"><i class="fa fa-check"></i><b>5.3.3</b> Detour to S3/S4</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#more-information"><i class="fa fa-check"></i><b>5.4</b> More information</a><ul>
<li class="chapter" data-level="5.4.1" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#checking-for-help-for-any-function"><i class="fa fa-check"></i><b>5.4.1</b> Checking for help for any function!</a></li>
</ul></li>
<li class="chapter" data-level="5.5" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#grammer-of-graphics-ggplot2"><i class="fa fa-check"></i><b>5.5</b> Grammer of Graphics (ggplot2)</a><ul>
<li class="chapter" data-level="5.5.1" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#what-is-ggplot2"><i class="fa fa-check"></i><b>5.5.1</b> What is ggplot2?</a></li>
<li class="chapter" data-level="5.5.2" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#principles-of-ggplot2"><i class="fa fa-check"></i><b>5.5.2</b> Principles of ggplot2</a></li>
</ul></li>
<li class="chapter" data-level="5.6" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#reference"><i class="fa fa-check"></i><b>5.6</b> Reference</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="identifying-cell-populations.html"><a href="identifying-cell-populations.html"><i class="fa fa-check"></i><b>6</b> Identifying Cell Populations</a><ul>
<li class="chapter" data-level="6.1" data-path="identifying-cell-populations.html"><a href="identifying-cell-populations.html#google-slides"><i class="fa fa-check"></i><b>6.1</b> Google Slides</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html"><i class="fa fa-check"></i><b>7</b> Feature Selection and Cluster Analysis</a><ul>
<li class="chapter" data-level="7.1" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#abstract"><i class="fa fa-check"></i><b>7.1</b> Abstract</a></li>
<li class="chapter" data-level="7.2" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#seurat-tutorial-redo"><i class="fa fa-check"></i><b>7.2</b> Seurat Tutorial Redo</a><ul>
<li class="chapter" data-level="7.2.1" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#preprocessing-steps"><i class="fa fa-check"></i><b>7.2.1</b> Preprocessing Steps</a></li>
<li class="chapter" data-level="7.2.2" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#start-of-identifying-cell-types"><i class="fa fa-check"></i><b>7.2.2</b> Start of Identifying Cell Types</a></li>
<li class="chapter" data-level="7.2.3" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#run-non-linear-dimensional-reduction-umaptsne"><i class="fa fa-check"></i><b>7.2.3</b> Run non-linear dimensional reduction (UMAP/tSNE)</a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#feature-selection"><i class="fa fa-check"></i><b>7.3</b> Feature Selection</a><ul>
<li class="chapter" data-level="7.3.1" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#differential-expression-analysis"><i class="fa fa-check"></i><b>7.3.1</b> Differential Expression Analysis</a></li>
<li class="chapter" data-level="7.3.2" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#check-clusters"><i class="fa fa-check"></i><b>7.3.2</b> Check Clusters</a></li>
<li class="chapter" data-level="7.3.3" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#view-entire-object-structure"><i class="fa fa-check"></i><b>7.3.3</b> View Entire Object Structure</a></li>
<li class="chapter" data-level="7.3.4" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#practice-visualizingembedding"><i class="fa fa-check"></i><b>7.3.4</b> Practice Visualizing/Embedding</a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#other-options-for-analysis"><i class="fa fa-check"></i><b>7.4</b> Other Options For Analysis</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="single-cell-resources.html"><a href="single-cell-resources.html"><i class="fa fa-check"></i><b>8</b> Single Cell Resources</a><ul>
<li class="chapter" data-level="8.1" data-path="single-cell-resources.html"><a href="single-cell-resources.html#comprehensive-list-of-single-cell-resources"><i class="fa fa-check"></i><b>8.1</b> Comprehensive list of single-cell resources</a></li>
<li class="chapter" data-level="8.2" data-path="single-cell-resources.html"><a href="single-cell-resources.html#computational-packages-for-single-cell-analysis"><i class="fa fa-check"></i><b>8.2</b> Computational packages for single-cell analysis</a></li>
<li class="chapter" data-level="8.3" data-path="single-cell-resources.html"><a href="single-cell-resources.html#elife-commentary-on-the-human-cell-atlas"><i class="fa fa-check"></i><b>8.3</b> eLife Commentary on the Human Cell Atlas</a></li>
<li class="chapter" data-level="8.4" data-path="single-cell-resources.html"><a href="single-cell-resources.html#online-courses"><i class="fa fa-check"></i><b>8.4</b> Online courses</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">ANALYSIS OF SINGLE CELL RNA-SEQ DATA</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="header">
<h1 class="title">ANALYSIS OF SINGLE CELL RNA-SEQ DATA</h1>
<p class="author"><em>Orr Ashenberg</em></p>
<p class="author"><em>Dana Silverbush</em></p>
<p class="author"><em>Kirk Gosik</em></p>
<p class="date"><em>03/23/2020 - 03/27/2020</em></p>
</div>
<div id="introduction" class="section level1">
<h1><span class="header-section-number">1</span> Introduction</h1>
<div id="course-overview" class="section level2">
<h2><span class="header-section-number">1.1</span> COURSE OVERVIEW</h2>
<p>In recent years single cell RNA-seq (scRNA-seq) has become widely used for transcriptome analysis in many areas of biology. In contrast to bulk RNA-seq, scRNA-seq provides quantitative measurements of the expression of every gene in a single cell. However, to analyze scRNA-seq data, novel methods are required and some of the underlying assumptions for the methods developed for bulk RNA-seq experiments are no longer valid. In this course we will cover all steps of the scRNA-seq processing, starting from the raw reads coming off the sequencer. The course includes common analysis strategies, using state-of-the-art methods and we also discuss the central biological questions that can be addressed using scRNA-seq.</p>
</div>
<div id="targeted-audience-assumed-background" class="section level2">
<h2><span class="header-section-number">1.2</span> TARGETED AUDIENCE &amp; ASSUMED BACKGROUND</h2>
<p>This course is aimed at researchers and technical workers who are or will be analyzing scRNA-seq data. The material is suitable both for experimentalists who want to learn more about data-analysis as well as computational biologists who want to learn about scRNASeq methods. Examples demonstrated in this course can be applied to any experimental protocol or biological system.</p>
<p>The requirements for this course are:
1. Working knowledge of unix (managing files, running programs)
2. Programming experience in R (writing a function, basic I/O operations, variable types, using packages). Bioconductor experience is a plus.
3. Familiarity with next-generation sequencing data and its analyses (using alignment and quantification tools for bulk sequencing data)</p>
</div>
<div id="course-format" class="section level2">
<h2><span class="header-section-number">1.3</span> COURSE FORMAT</h2>
<p>The course will be delivered over the course of five days. Each day will include a lecture and laboratory component. The lecture will introduce the topics of discussion and the laboratory sessions will be focused on practical hands-on analysis of scRNA-seq data. These sessions will involve a combination of both mirroring exercises with the instructor to demonstrate a skill as well as applying these skills on your own to complete individual exercises. After and during each exercise, interpretation of results will be discussed as a group. Computing will be done using a combination of tools installed on the attendees laptop computer and web resources accessed via web browser.</p>
</div>
<div id="getting-started" class="section level2">
<h2><span class="header-section-number">1.4</span> Getting Started</h2>
</div>
<div id="session-content" class="section level2">
<h2><span class="header-section-number">1.5</span> SESSION CONTENT</h2>
<div id="monday" class="section level3">
<h3><span class="header-section-number">1.5.1</span> Monday</h3>
<ul>
<li>Classes from 08:00 to 16:00 (lunch break-1 hr, 40 min of total coffee breaks)</li>
<li>Shared Google doc - course notes, ideas/questions/challenges/interesting topics you would like to explore.</li>
</ul>
<div id="lecture-1-scrna-seq-experimental-design-orr" class="section level4">
<h4><span class="header-section-number">1.5.1.1</span> Lecture 1 – scRNA-Seq experimental design (Orr)</h4>
<ul>
<li>Overview of course</li>
<li>General introduction: cell atlas overviews</li>
<li>Comparison of Bulk and single cell RNA-Seq</li>
<li>Overview of available scRNA-seq technologies (10x) and experimental protocols</li>
<li>scRNA-Seq experimental design and analysis workflow?</li>
</ul>
</div>
<div id="lab-1-understanding-sequencing-raw-data-downloading-docker-if-not-done-already-kirk" class="section level4">
<h4><span class="header-section-number">1.5.1.2</span> Lab 1 – Understanding sequencing raw data, downloading Docker if not done already (Kirk)</h4>
<p>Lab based around data wrangling from public data repositories: get data from 10x website, single cell portal, from GEO (fastqs, counts)
- Shell and Unix commands to navigate directories, create folders, open files
- Raw file formats</p>
</div>
<div id="lecture-2-3---intro-to-data-processing-from-bcl-file-to-bam-file" class="section level4">
<h4><span class="header-section-number">1.5.1.3</span> Lecture 2-3 - Intro to Data processing: from bcl file to bam file</h4>
<p><strong>Transcriptome quantification: from bam file to counts (Dana)</strong></p>
<ul>
<li>scRNA-Seq processing workflow starting with choice of sequencer (NextSeq, HiSeq, MiSeq) / barcode swapping and bcl files</li>
<li>Overview of Popular tools and algorithms</li>
<li>Common single-cell analyses and interpretation</li>
<li>Sequencing data: alignment and quality control</li>
<li>Looking at cool things in alignment like where reads are, mutations, splicing</li>
<li>Read &amp; UMI counting (Kallisto alignment-free pseudocounts as well), how RSEM works (length dependence, sequencing depth, multimapping reads), CellRanger (dropest), bustools</li>
<li>10x barcode structure and links to Perturb-seq</li>
<li>Gene length &amp; coverage</li>
<li>Gene expression units (count data Smart-seq2 counts or 10x UMIs vs expression data)</li>
</ul>
</div>
<div id="lab-2-processing-raw-scrna-seq-data-dana-docker-setup-kirk" class="section level4">
<h4><span class="header-section-number">1.5.1.4</span> Lab 2 – Processing raw scRNA-Seq data (Dana), Docker setup (Kirk)</h4>
<ul>
<li>Data outputs from different scRNAseq technologies (10x, Smart-seq2) - process both?</li>
<li>Demultiplexing sequencing data</li>
<li>Read Quality Control (CellRanger, dropEst, fastqc)</li>
<li>Run bowtie2 on 2 wells to demonstrate alignment</li>
<li>Read alignment and visualization (kallisto, RSEM, Igviewer)</li>
<li>Demultiplexing</li>
<li>FastQC</li>
<li>Align (STAR/TOPHAT/Kallisto)</li>
<li>IGViewer - what do we want here? I use it for mutation detections, copying sequences, searching for alternative splicing.</li>
</ul>
</div>
<div id="flash-talks-1.5-hr-break-into-2-groups-of-13" class="section level4">
<h4><span class="header-section-number">1.5.1.5</span> Flash talks (1.5 hr, break into 2 groups of 13)</h4>
<p>1 slide advertising or summarizing the poster. So you can introduce yourselves and we can get to know each other. No questions, 2 minutes. Two sessions, 15 people each.</p>
</div>
</div>
<div id="tuesday-classes-from-0800-to-1600" class="section level3">
<h3><span class="header-section-number">1.5.2</span> Tuesday – Classes from 08:00 to 16:00</h3>
<div id="lab-3---introduction-to-r-kirk" class="section level4">
<h4><span class="header-section-number">1.5.2.1</span> Lab 3 - Introduction to R (Kirk)</h4>
<ul>
<li>Some R overview slides, <a href="https://r4ds.had.co.nz/" class="uri">https://r4ds.had.co.nz/</a></li>
<li>Installing packages</li>
<li>Data-types</li>
<li>Data manipulation, slicing</li>
<li>Strings manipulations</li>
<li>Introducing object oriented programming / S4 objects</li>
<li>Visualization tools</li>
<li>Bonus create FeaturePlot from Seurat in base ggplot</li>
<li>Bonus: run RSEM on Dana’s bam files if you are bored</li>
</ul>
</div>
<div id="lecture-2-3---expression-qc-normalisation-and-gene-level-batch-correction-orr" class="section level4">
<h4><span class="header-section-number">1.5.2.2</span> Lecture 2-3 - Expression QC, normalisation and gene-level batch correction (Orr)</h4>
<ul>
<li>What CellRanger does for quality filtering</li>
<li>PBMC data</li>
<li>Normalisation methods <a href="https://www.nature.com/articles/nmeth.4292" class="uri">https://www.nature.com/articles/nmeth.4292</a></li>
<li>Doublets, empty droplets, CellBender</li>
<li>Barcode swapping</li>
<li>Regression with technical covariates</li>
<li>What about imputation?</li>
</ul>
</div>
<div id="lab-4-data-wrangling-for-scrnaseq-data-dana" class="section level4">
<h4><span class="header-section-number">1.5.2.3</span> Lab 4 – Data wrangling for scRNAseq data (Dana)</h4>
<ul>
<li>Data structures and file formats for single-cell data</li>
<li>Quality control of cells and genes (doublets, ambient, empty drops)</li>
<li>Data exploration: violin plots…</li>
<li>Introducing Seurat object
<ul>
<li>Genes</li>
<li>House keeping genes</li>
<li>Mitochondrial genes (never used these ones)</li>
<li>Filter - Do we remove both cells and genes here?</li>
</ul></li>
<li>Normalize (introduce more options, other than log transform?)</li>
<li>Find variable genes (Is it a first reduction? Why the binning?)</li>
<li>Scaling</li>
<li>Regression</li>
<li>Heatmap of desired genes?</li>
<li>Signatures?</li>
<li>Bonus - imputation (magic? One of the two Gocken recommended?)</li>
</ul>
</div>
<div id="flash-talks-1.5-hr-break-into-2-groups-of-13-1" class="section level4">
<h4><span class="header-section-number">1.5.2.4</span> Flash talks (1.5 hr, break into 2 groups of 13)</h4>
<p>1 slide advertising or summarizing the poster. So you can introduce yourselves and we can get to know each other. No questions, 2 minutes. Two sessions, 15 people each.</p>
</div>
</div>
<div id="wednesday-classes-from-0800-to-1600" class="section level3">
<h3><span class="header-section-number">1.5.3</span> Wednesday – Classes from 08:00 to 16:00</h3>
<div id="lecture-4-may-start-late-tuesday---identifying-cell-populations-kirk" class="section level4">
<h4><span class="header-section-number">1.5.3.1</span> Lecture 4 (may start late Tuesday) - Identifying cell populations (Kirk)</h4>
<ul>
<li>Feature selection</li>
<li>Dimensionality reduction</li>
<li>Clustering and assigning identity (Louvain, NMF, topic models, variational autoencoder)</li>
<li>Differential expression tests</li>
</ul>
</div>
<div id="lab-5-feature-selection-clustering-analysis-kirk" class="section level4">
<h4><span class="header-section-number">1.5.3.2</span> Lab 5 – Feature selection &amp; Clustering analysis (Kirk)</h4>
<ul>
<li>Parameters and clustering</li>
<li>Comparison of feature selection methods</li>
</ul>
</div>
<div id="lecture-5---batch-effects-correction-orr" class="section level4">
<h4><span class="header-section-number">1.5.3.3</span> Lecture 5 - Batch effects correction (Orr)</h4>
<ul>
<li>Batch correction methods (regress out batch, scaling within batch, Seurat v3, MNN, Liger, Harmony, scvi, scgen)</li>
<li>Evaluation methods for batch correction (ARI, average silhouette width, kBET…)</li>
</ul>
</div>
<div id="lab-6---correcting-batch-effects-orr" class="section level4">
<h4><span class="header-section-number">1.5.3.4</span> Lab 6 - Correcting batch effects (Orr)</h4>
<ul>
<li>Comparison of batch correction methods, Seurat pancreas</li>
<li>Use Seurat Wrappers?</li>
</ul>
</div>
</div>
<div id="thursday-classes-from-0800-to-1600" class="section level3">
<h3><span class="header-section-number">1.5.4</span> Thursday – Classes from 08:00 to 16:00</h3>
<p><strong>Deciding on discussion topics for next day based on shared google doc.</strong></p>
<div id="lecture-6---advanced-topics-kirk" class="section level4">
<h4><span class="header-section-number">1.5.4.1</span> Lecture 6 - Advanced topics (Kirk)</h4>
<ul>
<li>Waddington Landscape</li>
<li>Pseudotime inference</li>
<li>Differential expression through pseudotime</li>
<li>Deep learning or spatial data depending on questionnaire (20ish min, autoencoder as nonlinear dimension reduction, scvi, what questions to ask to evaluate whether a more advanced model helps, how to decide it’s safe to use a method, tradeoffs between method complexity and interpretability)</li>
</ul>
</div>
<div id="lab-7---functional-and-pseudotime-analysis-orr" class="section level4">
<h4><span class="header-section-number">1.5.4.2</span> Lab 7 - Functional and Pseudotime analysis (Orr)</h4>
<ul>
<li>Popular tools and packages for functional analysis (<a href="https://github.com/dynverse/dynmethods#list-of-included-methods" class="uri">https://github.com/dynverse/dynmethods#list-of-included-methods</a>)</li>
<li>Review concepts from papers</li>
<li>Comparison of pseudotime methods</li>
</ul>
</div>
<div id="lecture-7---single-cell-multi-omic-technologies-dana" class="section level4">
<h4><span class="header-section-number">1.5.4.3</span> Lecture 7 - Single-cell multi-omic technologies (Dana)</h4>
<ul>
<li>Introduction to other omic data types</li>
<li>Integrating scRNA-seq with other single-cell modalities (CITE, Perturb, ATAC, methylation…)</li>
</ul>
</div>
<div id="lab-8---analysis-of-cite-seq-scatac-seq-orr" class="section level4">
<h4><span class="header-section-number">1.5.4.4</span> Lab 8 - Analysis of CITE-seq, scATAC-seq (Orr)</h4>
<ul>
<li><a href="https://github.com/Hoohm/CITE-seq-Count" class="uri">https://github.com/Hoohm/CITE-seq-Count</a></li>
<li><a href="https://cite-seq.com/eccite-seq/" class="uri">https://cite-seq.com/eccite-seq/</a></li>
<li><a href="https://support.10xgenomics.com/single-cell-vdj/index/doc/technical-note-assay-scheme-and-configuration-of-chromium-single-cell-vdj-libraries" class="uri">https://support.10xgenomics.com/single-cell-vdj/index/doc/technical-note-assay-scheme-and-configuration-of-chromium-single-cell-vdj-libraries</a></li>
<li><a href="https://satijalab.org/seurat/multimodal_vignette.html" class="uri">https://satijalab.org/seurat/multimodal_vignette.html</a></li>
<li><a href="https://www.bioconductor.org/packages/devel/bioc/vignettes/cicero/inst/doc/website.html" class="uri">https://www.bioconductor.org/packages/devel/bioc/vignettes/cicero/inst/doc/website.html</a></li>
</ul>
</div>
</div>
<div id="friday-classes-from-0800-to-1600" class="section level3">
<h3><span class="header-section-number">1.5.5</span> Friday – Classes from 08:00 to 16:00</h3>
<p>Small group discussion on selected topics through hangouts.</p>
<div id="lab-10---small-dataset-for-analysis-and-office-hours-focused-on-select-topics-dana" class="section level4">
<h4><span class="header-section-number">1.5.5.1</span> Lab 10 - small dataset for analysis and office hours focused on select topics (Dana)</h4>
<p>For project on last day (plan for whole day), Dana will prepare datasets for 3 or more ish mut glioma tumors that they will download beforehand. The datasets may need to be subsampled to save time. Can do pseudotime, can run scvi, nmf. Groups of 3 students.</p>
</div>
<div id="group-presentations" class="section level4">
<h4><span class="header-section-number">1.5.5.2</span> Group presentations</h4>
<p>Review, Questions and Answers</p>

</div>
</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>

<a href="lab-template.html" class="navigation navigation-next navigation-unique" aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["2020_scWorkshop.pdf", "2020_scWorkshop.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
